### DISCUSSION TITLE
[V3] active link class using wire:navigate in persist mode

### DISCUSSION THREAD

**Discussion by ev3rlost** [livewire/livewire#6043] [1. Help] [answered]
How do we implement an active class on a link while using `wire:navigate` and having the whole `<nav></nav>` as a Livewire component in **persist** mode?

Simply putting `class="{{ (request()->is('dashboard')) ? 'active' : '' }}"` on the link doesn't seem to work.

I'm importing the component in 'resources/views/components/layouts/app.blade.php' like this:

```
<livewire:layout.nav />
```

The code for the component:

```
<nav x-persist="nav">
    <a href="/dashboard" class="{{ (request()->is('dashboard')) ? 'active' : '' }}">
        Dashboard
    </a>
</nav>
```

Is there a solution? Am I thinking in the wrong direction? Should I use Alpine to determine the URL and pop an active class on the link?

**Comment by nleibbrandt** [comment]
Maybe you can use the name of the livewire component, $this->getName() === 'dashboard' ?

**Reply by ev3rlost** [reply]
Could you give a code example please?

**Comment by cainydev** [comment]
I am struggeling with the exact same thing. I have a navigation that I want to persist. The problem is that I now have to send a livewire event to that component to manually rerender the new active button.

I would just not use persist but then all the submenus that are expanded in the moment of navigating are closed again (alpine data is reset)

I even tried to use the `wire:navigated` event but it seems that the url is still the old one when this event is fired. I feel like that needs some work. Maybe a `wire:navigating` and a `wire:navigated` event.

```html
x-on:wire:navigated.document="active = window.location.pathname.split('/').pop() == page"
```
gives me the old page-id..

**Reply by cainydev** [reply]
As a workaround, you could try this:

1. instead of using `wire:navigate`, use `wire:click` to call this function:

```php
public function goto(string $target){
    $this->dispatch('navigate-to', target: $target);
    return $this->redirect(route($target), navigate: true);
}
```

2. Listen for the event in your navigation item:

```php
#[On('navigate-to')]
public function onNavigated(string $target) {
    $this->active = ($target == $this->entry->name);
}
```

Sorry for any typos or errors, this is just from my head.

Also, this could get costly when you have dozens of menu entries, for example in a dynamic navigation. Performance wise it would probably be smart to fire a dynamic event that only targets the old navigation item and the new one.

**Reply by ev3rlost** [reply]
Kinda defeats the purpose of having `wire:navigate` at all. Nice workaround, although too much of a hassle for a fairly simple task. I think this needs to be adressed to the maintainers as a bug and needs to be fixed, as it's something people will be using a lot when creating an app.

**Comment by donkooijman** [comment]
@ev3rlost checking against the route name works for me. This does require you to name your routes in your routes file. See example below:

```
@foreach(['home','statistics', 'settings'] as $route)
    <li>
        <a class="{{ request()->routeIs($route) ? 'text-black' : 'text-blue-500' }}"
           wire:navigate
           href="{{ route($route) }}">
            {{ $route }}
        </a>
    </li>
@endforeach
```

**Reply by cainydev** [reply]
Is that wrapped in a `@persist` ?

**Reply by ev3rlost** [reply]
As [cainytheslave](https://github.com/cainytheslave) said, your code breaks if we put the code inside `@persist` or add `x-persist` on the component parent div.

**Reply by donkooijman** [reply]
@ev3rlost @cainytheslave This is not wrapped in @persist, I was assuming you doing that for the functionality but it's not required. It makes sense to me that it's not wokring since the persist section is not ran again. 

By the way, I'm curious to know, why did you opt to persist this particular navigation?

**Reply by ev3rlost** [reply]
My header and navigation have functionalities and don't have to be reloaded every time I navigate to a different page.

**Reply by donkooijman** [reply]
Is it possible to wrap those sections that don't need to be reloaded in persist and keep the checking of current route outside of persist?

**Reply by cainydev** [reply]
For me it's a use-case where I have a large navigation tree on the left (about 150 items) and only 2 small style changes are necessary when navigating so I thought this was really unnessecary to reload the whole thing.

**Reply by donkooijman** [reply]
By using persist, you instruct Alpine to not touch that section (just keep what is already there). Any dynamic changes inside of the persist will not be updated, which makes sense because you tell it not to.

**Reply by ev3rlost** [reply]
Show and hide functionality for the navigation for example. The state needs to persist throughout page navigations.

**Reply by cainydev** [reply]
@ev3rlost Maybe just use alpine magic `$persist` for persisting the alpine state..

**Reply by addtobag** [reply]
@ev3rlost did you manage to resolve this? I'm looking for something similar.

**Comment by JeffGepiga** [comment]
Try this. it works for me.

In main lay-out

``` 
<body>
<script data-navigate-once>
   window.showActive = function(active_link=null) {
              active_link = active_link??document.URL.split(/[?#]/)[0];
              document.querySelectorAll('#sidebar-menu a.nav-link').forEach(function(item) {
                if (active_link==item.getAttribute("href")) {
                    item.classList.add('active');    
                }else{
                    item.classList.remove('active')
                }
            });
        };
        document.querySelectorAll('#sidebar-menu a.nav-link').forEach(element => element.addEventListener('click', e => {
            showActive(element.getAttribute("href"));
        }));
        window.showActive();
    </script>
 </body>
```

In Sidebar with persist
```
<div x-persist="sidebar" class="collapse navbar-collapse" id="sidebar-menu" >
            <ul class="navbar-nav pt-lg-3">
              <li class="nav-item" >
                <a class="nav-link" href="{{url('dashboard')}}" wire:navigate>
                  <span class="nav-link-icon d-md-none d-lg-inline-block">
                   <i class="ti ti-dashboard fs-2"></i>
                  </span>
                  <span class="nav-link-title">
                    Dashboard
                  </span>
                </a>
              </li>
              <li class="nav-item">
                <a class="nav-link" href="{{url('training')}}" wire:navigate>
                  <span class="nav-link-icon d-md-none d-lg-inline-block">
                   <i class="ti ti-dashboard fs-2"></i>
                  </span>
                  <span class="nav-link-title">
                    Training
                  </span>
                </a>
           </li>
    </ul>
</div
```

**Comment by cainydev** [comment]
I want to suggest another nice solution using very little alpine logic.

When you persist your navigation, make your navigation links (that should be active/not active after wire:navigate) like this. You might need to adjust your URL parser but thats a minor inconvenience.. I also tried to use `x-on:wire:navigated.document` but without success.

```blade
@props(['name'])

<a href="{{ route($name) }}" 
   wire:navigate
   x-data="{ active: false, u() { this.active = (window.location.href.includes('{{ $name }}') ) }}"
   x-init="u()"
   x-on:alpine:navigated.document="$nextTick(u)"
   x-bind:class="active && 'text-red-500'">

{{ $page->title }}

</a>
```

**Comment by JSFTI** [comment]
I noticed that the Livewire documentation sidebar has a similar feature.

This is how they did it.
```
<script data-navigate-once>
document.addEventListener('alpine:init', () => {
    let state = Alpine.reactive({ path: window.location.pathname })

    document.addEventListener('livewire:navigated', () => {
        queueMicrotask(() => {
            state.path = window.location.pathname
        })
    })

    Alpine.magic('current', (el) => (expected = '') => {
        let strip = (subject) => subject.replace(/^\/|\/$/g, '')

        return strip(state.path) === strip(expected)
    })
})
</script>
```

Then, to check active route, they use 
```
x-bind:class="$current('/') ? 'active-class' : 'not-active-class'"
```
Example:
```
@persist('navigation')
    <div>
        <a 
            href="/" wire:navigate 
            x-bind:class="$current('/') ? 'text-red-500' : ''"
        >
            Root
        </a>
        <a 
            href="/other-page" wire:navigate
            x-bind:class="$current('/other-page') ? 'text-red-500' : ''"
        >
            Other Page
        </a>
    </div>
@endpersist
<script data-navigate-once>
    document.addEventListener('alpine:init', () => {
        let state = Alpine.reactive({ path: window.location.pathname })
    
        document.addEventListener('livewire:navigated', () => {
            queueMicrotask(() => {
                state.path = window.location.pathname
            })
        })
    
        Alpine.magic('current', (el) => (expected = '') => {
            let strip = (subject) => subject.replace(/^\/|\/$/g, '')
            return strip(state.path) === strip(expected)
        })
    }) 
</script>
```

**Reply by Joemires** [reply]
Nice one @JSFTI  🖤
I made mine this way tho
```html
<script data-navigate-once>
    document.addEventListener('alpine:init', () => {
        let state = Alpine.reactive({ path: window.location.pathname })

        document.addEventListener('livewire:navigated', () => {
            queueMicrotask(() => {
                state.path = window.location.pathname
            })
        })

        let strip = (subject) => subject.replace(/^\/|\/$/g, '')

        if (! String.prototype.startsWith) {
                String.prototype.startsWith = function(searchString, position){
                position = position || 0;
                return this.substr(position, searchString.length) === searchString;
            };
        }

        Alpine.magic('current', (el) => (expected = '') => {
            let regexMatch = strip(expected).substr(-1) === '*';
            return regexMatch ? strip(state.path).startsWith(strip(expected.replace(/.$/, ""))) : strip(state.path) === strip(expected)
        })
    })
</script>
```

<br>

With this method, I can now have something like

```
x-bind:class="$current('/') ? 'active-class' : 'not-active-class'"
```
OR
```
x-bind:class="$current('account/*') ? 'active-class' : 'not-active-class'"
```

**Comment by jdecode** [comment]
Note : This has NOT been tested in persist mode.

A more "Laravel" solution that is posted on Laracasts Forum is to use a custom Blade directive.
https://laracasts.com/discuss/channels/guides/blade-directive-to-add-a-class-to-active-url

While giving the demo of Livewire 3.x the very first time, @calebporzio used a directive `@active('dashboard')` (in the very first code exercise) which looked pretty neat, and I created a similar version for myself and it has worked well so far.

```php
Blade::directive('active', function ($route) {
    return "<?php echo Route::currentRouteNamed($route)
        ?
        ' class=\"active\" '
        :
        ' class=\"inactive\" '
    ; ?>";
});
```

You may want to use multiple classes to differentiate the active state - but the core logic keeps the resultant HTML very neat.

Sample usage: 
```php
<a wire:navigate href="{{ route('dashboard') }}" @active('dashboard')>
```

Laravel docs to create custom directive : https://laravel.com/docs/10.x/blade#extending-blade

**Reply by cainydev** [reply]
Well thats nice but doesnt work in persist mode.

**Comment by MrCrayon** [comment]
IMHO the easiest way to pass a parameter to a persisted template is to use `Alpine.$store`.

For the original question would be something like:
```blade
<script>
  document.addEventListener('livewire:navigated', () => {
    Alpine.store('isDashboard', @js(request()->is('dashboard')));
  });
</script>
```
and then
```blade
<nav x-persist="nav">
    <a href="/dashboard" :class="$store.isDashboard ? 'active' : ''">
        Dashboard
    </a>
</nav>
```

**Comment by oscaralderete** [comment]
These days I'm preparing a simple Laravel Livewire demo for a potential client, I faced the same problem discussed on this thread and here's my approach to add a distinctive CSS class name to current url in a Laravel Livewire navigation bar.

1. Define a micro Alpinejs app:
```
<nav x-data="{ active: '{{ str_replace(url()->query(''), '', url()->current()) }}' }">
	// links will be here
</nav>
```
Livewire returns the current base url with: url()->query('') and the full url with: url()->current(), as result data.active will be '/contact', '/about', etc. In load time, home page is an empty string.

2. Then we populate the links, besides the href attribute each of them has a click event that assigns to data.active its corresponding path. To assign the CSS class active we only need to compare if data.active matchs with the current path:

```
<a href="/" @click="active='/'" class="active" :class="{ 'active': active == '/' || active == '' }" wire:navigate>Home</a>
<a href="/about" @click="active='/about'" :class="{ 'active': active == '/about' }" wire:navigate>About Us</a>
```

note that Home requires a double comparison rule, this is because the navigation click event assigns '/' to data.active, but on load time its an empty string.

3. That's all! At the end we have something like:
```
<nav x-data="{ active: '{{ str_replace(url()->query(''), '', url()->current()) }}' }">
	<a href="/" @click="active='/'" class="active" :class="{ 'active': active == '/' || active == '' }" wire:navigate>Home</a>
	<a href="/about" @click="active='/about'" :class="{ 'active': active == '/about' }" wire:navigate>About Us</a>
	<a href="/posts" @click="active='/posts'" :class="{ 'active': active == '/posts' }" wire:navigate>Posts</a>
	<a href="/users" @click="active='/users'" :class="{ 'active': active == '/users' }" wire:navigate>Users</a>
	<a href="/contact" @click="active='/contact'" :class="{ 'active': active == '/contact' }"
     wire:navigate>Contact</a>
</nav>
```

To view it in action I uploaded a video to my YT channel, watch it here:
https://www.youtube.com/watch?v=Rqg0h4E3u_c

**Comment by joshhanley** [comment, marked as answer]
We're in the process of investigating all wire:navigate issues and I have tested out your scenario. I believe this is now possible using `wire:current` https://livewire.laravel.com/docs/wire-current, so I'm going to close this. If you're still experiencing problems, please open a new discussion. Thanks!
